extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/LDAPGuacamoleProperties.java

```diff
@@ -185,16 +185,6 @@ private LDAPGuacamoleProperties() {}
 
     };
 
-    /**
-     * Maximum number of referral hops to follow.
-     */
-    public static final IntegerGuacamoleProperty LDAP_MAX_REFERRAL_HOPS = new IntegerGuacamoleProperty() {
-
-        @Override
-        public String getName() { return "ldap-max-referral-hops"; }
```

Comment:
Is this not supported by the Apache Directory API?

Reply:
See my comment above - as far as I can tell, no, I have not found a place in the Apache Directory API that lets you control number of referral hops.  I've looked a couple more times to make sure and am not seeing it...

Reply:
Hm... OK. Well, assuming that this is impossible to replicate with the Apache Directory API, then I have no issue with its removal. Who am I to argue with reality. ;)

It's fine to remove support for a property, but we'll need to follow normal deprecation procedures. Ideally, this would mean allowing the property to continue functioning as normal, but failing that we at least need to warn in the logs when the property is used, notifying that the property is deprecated and no longer has any effect.

For example, after `auth-provider` and `lib-directory` were deprecated:

https://github.com/apache/guacamole-client/blob/027fa1d85e72352e3b9c634ae48ec5c0315b9672/guacamole/src/main/java/org/glyptodon/guacamole/net/basic/extension/ExtensionModule.java#L162-L165

Or with the deprecation of the `mysql-disallow-*` properties:

https://github.com/apache/guacamole-client/blob/162f4db386c3a612e8a1189e92b7f789d51945d8/extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-mysql/src/main/java/net/sourceforge/guacamole/net/auth/mysql/MySQLEnvironment.java#L123-L127

Reply:
> Hm... OK. Well, assuming that this is impossible to replicate with the Apache Directory API, then I have no issue with its removal. Who am I to argue with reality. ;)

Yeah...it is a shame, but I do think the benefits of moving to the Apache API outweigh keeping this one feature.

I do need to do a little more research and just make sure there's not something I'm missing.  I'm also wondering how hard it would be and if it would be worth it to try to implement some sort of form of this inside the Guacamole LDAP extension.  It seems like maybe in the sections that process the search results there might be a way to keep track of how many times a referral has occurred and drop out if it goes beyond the limit.  But, it may be too complicated to be worth implementing.

Reply:
I wonder if the `getReferralLength()` method could be used to check this?  It's a little hard to tell because the documentation on the Apache API is still in its early stages, and it's hard to tell what `referralLength` means in this context:

http://directory.apache.org/api/gen-docs/latest/apidocs/org/apache/directory/api/ldap/model/message/ReferralImpl.html#getReferralLength--

Thoughts??

Reply:
Yeah, the documentation for `getReferralLength()` doesn't really tell us anything. Looking through the underlying implementation, the value returned is manually populated through some external mechanism via `setReferralLength()`:

https://github.com/apache/directory-ldap-api/blob/9f888e56bcc65c6a5efe7a49bf7f59d45c9fc935/ldap/codec/core/src/main/java/org/apache/directory/api/ldap/codec/api/LdapEncoder.java#L345-L353

The above block of code gradually increments a length counter taking bytes into account:

    referralLength += 1 + TLV.getNbBytes( ldapUrlBytes.length ) + ldapUrlBytes.length;

Whatever referral length is supposed to be here, it seems to be more related to the overall byte length of some packet, not the depth of the chain of referrals.

Reply:
> It seems like maybe in the sections that process the search results there might be a way to keep track of how many times a referral has occurred and drop out if it goes beyond the limit.

This seems important functionality to have. As LDAP search utilities and the old JLDAP API both provide a means of defining the maximum number of hops (and since unbounded following of pointers is generally dangerous behavior), I think we should be absolutely certain that this is impossible before removing the functionality.

In fact, assuming that the functionality is indeed critical, and assuming that it's impossible to implement with the Apache Directory API, it might instead be best to:

1. Keep the property around, but warn that the underlying LDAP implementation does not currently implement it.
2. Open an issue with the Apache Directory project for implementing such a limit.

The Apache Directory mailing lists may be a good resource if we can't figure out what we need from the docs alone.

Reply:
> Yeah, the documentation for getReferralLength() doesn't really tell us anything. Looking through the underlying implementation, the value returned is manually populated through some external mechanism via setReferralLength():

Yeah, I looked at that after I sent that response and came to the same conclusion.  So, that's not going to work.

> This seems important functionality to have.

Yeah, I'd have to agree.  Seems like you could get yourself into a situation that could end badly without it.

> The Apache Directory mailing lists may be a good resource if we can't figure out what we need from the docs alone.

Already signed up for the api list and posted a message, there, so we'll see if I get a response.  It isn't an overly-active list, so not sure.

I'm a little afraid that, if an obvious solution doesn't come along, I'm going to have to implement the entire result processing functionality in the extension - which might be worth it, and might end up being necessary, anyway, to deal with paged results, but I really like leaving the implementation to the API and not having to do it myself.

Reply:
> The Apache Directory mailing lists may be a good resource if we can't figure out what we need from the docs alone.

Well, this particular issue ends up taking a new direction, since, apparently the Apache LDAP API does not automatically follow referrals, and that has to be manually coded.  So, guess I get to manually code that - and while I'm at it, keep track of/check referral depth.